pkg/services/cosmosdb/cosmos-utils.go

```diff
@@ -178,3 +182,218 @@ func deleteDatabase(
+}
+
+// For sqlAllInOneManager, the real type of `instance.Details` is
+// `*sqlAllInOneInstanceDetails`, so type assertion must be changed.
+// Expect type assertion, this function is totally the same as previous one.
+func (s *sqlAllInOneManager) waitForReadLocationsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*sqlAllInOneInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := s.databaseAccountsClient
+
+	err := pollingUntilReadLocationsReady(
+		ctx,
+		resourceGroupName,
+		accountName,
+		databaseAccountClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+const succeeded = "succeeded"
+
+// For now, this method will return on either context is cancelled or
+// every region's state is "succeeded" in seven consecutive check.
+// The reason why we need seven consecutive check is that
+// the read region is created one by one, there is a small gap between
+// the finishment of previous creation and the start of the next creation.
+// By this check, we can detect gaps shorter than 1 mintue,
+// and report success within 70 seconds after completion.
+func pollingUntilReadLocationsReady(
+	ctx context.Context,
+	resourceGroupName string,
+	accountName string,
+	databaseAccountClient cosmosSDK.DatabaseAccountsClient,
+) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Seven comsecutive check is needed to report success.
+	const confirmNumberOfTimes = 7
+
+	ticker := time.NewTicker(time.Second * 10)
+	previousSucceededTimes := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			result, err := databaseAccountClient.Get(
+				childCtx,
+				resourceGroupName,
+				accountName,
+			)
+			if err != nil {
+				return err
+			}
+
+			//Check whether every read location's state is "Succeeded"
+			allSucceed := true
+			readLocations := *(result.DatabaseAccountProperties.ReadLocations)
+			for i := range readLocations {
+				state := *(readLocations[i].ProvisioningState)
+				if strings.ToLower(state) != succeeded {
+					allSucceed = false
+					previousSucceededTimes = 0
+					break
+				}
+			}
+			if allSucceed && previousSucceededTimes >= confirmNumberOfTimes {
+				return nil
+			} else if allSucceed {
+				previousSucceededTimes++
+			}
+		}
+	}
+}
+
+func validateReadLocations(
+	context string,
+	regions []string,
+) error {
+	for i := range regions {
+		region := regions[i]
+		if !allowedReadLocations[region] {
+			return service.NewValidationError(
+				fmt.Sprintf("%s.allowedReadLocations", context),
+				fmt.Sprintf("given region %s is not allowed", region),
+			)
+		}
+	}
+	return nil
+}
+
+// Allowed CosmosDB read locations, it is different from Azure regions.
+// We use a map here to record all allowed regions.
+var allowedReadLocations = map[string]bool{
```

Comment:
Is there an API we can fetch this from instead of hard coding the list?

Also, do customers typically have access to euap reagions?

Reply:
I have gone over the [spec](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2015-04-08) provided by CosmosDB team and didn't find any API which can provide the read regions list. So we have to hard code it temporarily. And most customers don't have access to euap regions, we may delete all euap regions from the list.